### PR TITLE
Feat/r2lpl lifelong replay

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## Summary

This PR adds the R2LPL-inspired lifelong replay workflow for mining closed-loop reproducer failures, repairing them with generated candidate trajectories, and training cumulative replay rounds from the selected repaired scene-target pairs.

The workflow is documented in `rlvr/autoresearch/README_r2lpl_lifelong.md`, including its relationship to the R2LPL paper, *Learning from Mistakes: Rollout-Retrieval Lifelong Policy Learning for Autonomous Driving*, and the implementation differences from the paper.

Key changes:

- Add fixed-stride direct reproducer chunk mining over scene lists, with route/frame discontinuity cuts, pose-speed/yaw guards, MPC tracking, clock-time progression, simulated neighbor history, and per-shard outputs.
- Add a single workflow runner that launches sharded mining, mines credit-window scenes, runs K-candidate repair generation, builds replay memory, and starts the training round from one config-driven entrypoint.
- Add multi-GPU sharding support for mining and repair while keeping shard outputs isolated to avoid concurrent writes to the same files.
- Preserve scene-local vehicle geometry from NPZs, and save repaired NPZs with the simulated scene context plus the selected repaired ego future.
- Filter repaired NPZ payloads to training schema fields so mixed input corpora do not leak unrelated numeric metadata into downstream batching.
- Add tests for chunk planning, discontinuity guards, yaw wrap handling, sharding, repair commands, repaired payloads, DDP launch cleanup, and multi-GPU behavior.

## Validation

- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/pytest rlvr/autoresearch/tools/test_r2lpl_lifelong_replay.py -q`
- `uvx pre-commit run --files rlvr/autoresearch/tools/build_avoiding_target.py rlvr/autoresearch/tools/mine_credit_window_scenes.py rlvr/autoresearch/tools/mine_direct_reproducer_chunks.py rlvr/autoresearch/tools/test_r2lpl_lifelong_replay.py`
- `uvx pre-commit run --files rlvr/autoresearch/README_r2lpl_lifelong.md`

## Notes

The workflow is designed for large scene lists: chunk starts are sampled by fixed stride, each shard writes to its own output directory, and short/discontinuous windows are skipped or cut before simulation results are emitted.

## Update — realized-only detection, single collision algorithm, hard-example signal

Follow-up commits refine detection semantics and unify collision geometry:

- **Two-phase detection (closed-loop realized vs open-loop predicted).** The closed-loop reproducer now detects rule violations from the **realized ego pose at each sim step only** — the first realized violation anchors the credit window `[offense − gap − width, offense − gap]`. The model's predicted trajectory is no longer used to trigger reproducer events. The open-loop **repair** side is unchanged: it accepts a candidate whose **full predicted trajectory** clears every rule. Documented in `README_r2lpl_lifelong.md` ("Detection Semantics").
- **Single collision-detection algorithm.** New `planner_metrics/vehicle_collision.py` is a numpy/torch shim over the canonical OBB primitives; the GUI scene builder (hand-rolled SAT), parked-vehicle merge (shapely), and viz/eval corner builders are re-pointed to it (proven numerically identical). Moving-collision labeling routes through the reward's gated definition (rear-end / low-speed suppression, honoring `--count_rear_end_collisions`).
- **`target_gt_disagreement` hard-example signal.** Each repaired row records mean/max L2 between the selected repaired target and the original logged GT target — distinct from realized `expert_disagreement`.
- **Fixes.** Repair forces the collision gates its labels depend on (fail-loud instead of silently accepting neutralized gates); DER replay anchor is set to each round's prior checkpoint; `base_sft` fails loudly when asked for `replay_der_coef` / `replay_loss_weight` it cannot honor.
- **Cleanup.** Renamed the misleading manifest field `collision_step` → `window_end_step`; removed dead legacy paths `extract_collision_scenes` and `run_segment` (superseded by the one-pass save / `run_segments_batched`).

Validation: full unit suite passes (reward-frozen numbers unchanged); ruff clean; the full workflow was exercised end-to-end on a converted reproducer corpus (mining → repair → replay memory → SFT checkpoint).## Update — rear-end labeling + Copilot/agent review fixes

- **Rear-ends kept but labeled.** Instead of a detect/suppress toggle, a detected moving-collision now carries a `rear_end_collision` tag (`_moving_collision_all_rear_end`: pure rear-end = no forward-hemisphere overlap). The tag propagates realized-row → credit-window `manifest.json` → mined credit-window rows, so a downstream stage can drop a rear-end if the need arises instead of losing it at detection. `--count_rear_end_collisions` is threaded through mining (`mine_direct_reproducer_chunks`) so it matches the repair side and the RAW-`col` trigger.
- **Stopped/moving classification fix.** `current_ego_neighbor_clearance` uses horizon=2 for moving/static (so `_stopped_neighbor_mask` can classify stopped neighbors instead of mislabeling them moving), degrades to the available future length for single-realized-step t0 gates, and aligns `slot_valid` to the horizon window (fixes a latent static-path IndexError).
- **Per-scene `ego_shape`** in repaired metadata (was leaking the batch-loop `npz_es` = last validation scene).
- **`conflict_detector`** returns `expert_disagreement_step=None` (not `0`) for non-localized summary-based conflicts, matching the clean-result contract.
- **Credit-window offense frame** now uses the authoritative mined `offense_frame_id` (pose-mode cursor overshoot can no longer record the wrong offense frame).
- Dropped dead `_source_scene_t0_moving_overlap`; softened the `vehicle_collision` docstring (float64 + inclusive-touch differ from the reward path); reconciled the README 15-frame-span vs 16-inclusive-file wording.

Validation: 161 unit tests pass; ruff clean; two independent review passes (correctness + integration) — the integration pass caught the tag-propagation gap, now fixed. 10k-frame realized-only mine on one route surfaced 35 forward moving-collisions + 1 road-border crossing (0 pure rear-ends).

### CODE REVIEW

File: rlvr/grpo_sft_trainer.py [hunk 1]
```diff
@@ -371,79 +394,122 @@ def _compute_sft_diffusion_loss(
...
+            total_der_loss += der_loss
+
+    def _avg_per_scene(value) -> torch.Tensor:
+        if isinstance(value, torch.Tensor):
+            return value / K
+        return torch.zeros(B, device=device)
+
+    def _weighted(value: torch.Tensor) -> torch.Tensor:
+        return (value * scene_loss_weights).sum() / max(B, 1)
+
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply rlvr/grpo_sft_trainer.py:454]: Intentional, not a bug — this is the standard SFT batch-mean. With the default `scene_loss_weights = ones(B)`, `_weighted` reduces to `per_scene.sum() / B`, i.e. a plain mean over the minibatch (the same aggregation the original SFT loss uses). The weights multiply is an add-on for selective/replay masking, and dividing by B (not `sum(weights)`) is deliberate: it keeps each *active* scene at a constant `1/B` gradient contribution, so masking out normal scenes does not silently rescale the effective learning rate on the surviving problem scenes. Dividing by `sum(weights)` would inflate per-scene gradients as more scenes are masked, which would change the tuned selective-training recipes. Leaving as-is.
